File: crates/nil-core/src/military/maneuver.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::continent::{Coord, Distance};
use crate::error::{Error, Result};
use crate::military::army::ArmyId;
use crate::military::unit::stats::speed::Speed;
use serde::{Deserialize, Serialize};
use std::ops::{Sub, SubAssign};
use uuid::Uuid;

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct Maneuver {
  id: ManeuverId,
  army: ArmyId,
  kind: ManeuverKind,
  direction: ManeuverDirection,
  origin: Coord,
  destination: Coord,
  state: ManeuverState,
}

impl Maneuver {
  pub(super) fn new(request: &ManeuverRequest) -> Result<(ManeuverId, Self)> {
    let distance = request.origin.distance(request.destination);
    if request.origin == request.destination || distance == 0u8 {
      return Err(Error::OriginIsDestination(request.origin));
    }

    let id = ManeuverId::new();
    let maneuver = Self {
      id,
      army: request.army,
      kind: request.kind,
      direction: ManeuverDirection::Going,
      origin: request.origin,
      destination: request.destination,
      state: ManeuverState::new(distance.into()),
    };

    Ok((id, maneuver))
  }

  #[inline]
  pub fn id(&self) -> ManeuverId {
    self.id
  }

  #[inline]
  pub fn army(&self) -> ArmyId {
    self.army
  }

  #[inline]
  pub fn kind(&self) -> ManeuverKind {
    self.kind
  }

  #[inline]
  pub fn direction(&self) -> ManeuverDirection {
    self.direction
  }

  #[inline]
  pub fn origin(&self) -> Coord {
    self.origin
  }

  #[inline]
  pub fn destination(&self) -> Coord {
    self.destination
  }

  /// Checks whether the maneuver's origin or destination matches the coord.
  #[inline]
  pub fn matches_coord(&self, coord: Coord) -> bool {
    coord == self.origin || coord == self.destination
  }
}

#[must_use]
#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash, Deserialize, Serialize)]
pub struct ManeuverId(Uuid);

impl ManeuverId {
  pub fn new() -> Self {
    Self(Uuid::new_v4())
  }
}

impl Default for ManeuverId {
  fn default() -> Self {
    Self::new()
  }
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum ManeuverKind {
  Attack,
  Support,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
#[serde(rename_all = "kebab-case")]
pub enum ManeuverDirection {
  Going,
  Returning,
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(tag = "kind", rename_all = "kebab-case")]
pub enum ManeuverState {
  Done,
  Pending { distance: ManeuverDistance },
}

impl ManeuverState {
  fn new(distance: ManeuverDistance) -> Self {
    Self::Pending { distance }
  }
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize)]
pub struct ManeuverDistance(f64);

impl Sub for ManeuverDistance {
  type Output = ManeuverDistance;

  fn sub(mut self, rhs: Self) -> Self::Output {
    self -= rhs;
    self
  }
}

impl Sub<Speed> for ManeuverDistance {
  type Output = ManeuverDistance;

  fn sub(mut self, rhs: Speed) -> Self::Output {
    self -= rhs;
    self
  }
}

impl SubAssign for ManeuverDistance {
  fn sub_assign(&mut self, rhs: Self) {
    *self = Self(self.0 - rhs.0);
  }
}

impl SubAssign<Speed> for ManeuverDistance {
  fn sub_assign(&mut self, rhs: Speed) {
    *self = Self(self.0 - f64::from(rhs));
  }
}

impl From<Distance> for ManeuverDistance {
  fn from(distance: Distance) -> Self {
    Self(f64::from(distance))
  }
}

#[derive(Clone, Debug, Deserialize, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ManeuverRequest {
  pub army: ArmyId,
  pub kind: ManeuverKind,
  pub origin: Coord,
  pub destination: Coord,
}

File: app/src/core/model/military/maneuver.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { go } from "@/router";
import * as commands from "@/commands";
import { getManeuver } from "@/commands";
import type { Option } from "@tb-dev/utils";
import { CoordImpl } from "@/core/model/continent/coord";
import { PublicCityImpl } from "@/core/model/city/public-city";
import { ManeuverHaulImpl } from "@/core/model/military/maneuver-haul";
import type {
  ArmyId,
  Maneuver,
  ManeuverDirection,
  ManeuverId,
  ManeuverKind,
  ManeuverState,
  Ruler,
} from "@tsukilabs/nil-bindings";

export class ManeuverImpl implements Readonly<Maneuver> {
  public readonly id: ManeuverId;
  public readonly origin: CoordImpl;
  public readonly destination: CoordImpl;
  public readonly army: ArmyId;
  public readonly kind: ManeuverKind;
  public readonly direction: ManeuverDirection;
  public readonly state: ManeuverState;
  public readonly speed: number;
  public readonly hauledResources: ManeuverHaulImpl | null;

  #armyOwner: Option<Ruler>;
  #originCity: Option<PublicCityImpl>;
  #destinationCity: Option<PublicCityImpl>;

  private constructor(maneuver: Maneuver) {
    this.id = maneuver.id;
    this.origin = CoordImpl.create(maneuver.origin);
    this.destination = CoordImpl.create(maneuver.destination);
    this.army = maneuver.army;
    this.kind = maneuver.kind;
    this.direction = maneuver.direction;
    this.state = maneuver.state;
    this.speed = maneuver.speed;

    this.hauledResources = maneuver.hauledResources ?
      ManeuverHaulImpl.create(maneuver.hauledResources) :
      null;
  }

  public isAttack() {
    return this.kind === "attack";
  }

  public isSupport() {
    return this.kind === "support";
  }

  public isGoing() {
    return this.direction === "going";
  }

  public isReturning() {
    return this.direction === "returning";
  }

  public isDone() {
    return this.state.kind === "done";
  }

  public isPending() {
    return this.state.kind === "pending";
  }

  public getPendingDistance() {
    if (this.state.kind === "pending") {
      return this.state.distance;
    }
    else {
      return 0;
    }
  }

  public async goToManeuverScene() {
    await go("maneuver", {
      params: { id: this.id },
    });
  }

  public hasHauledResources() {
    if (this.hauledResources) {
      return !this.hauledResources.isEmpty();
    }
    else {
      return false;
    }
  }

  public async getArmyOwner(): Promise<Ruler> {
    this.#armyOwner ??= await commands.getArmyOwner(this.army);
    return this.#armyOwner;
  }

  public async getCities() {
    if (!this.#originCity || !this.#destinationCity) {
      const cities = await PublicCityImpl.bulkLoad([this.origin, this.destination]);
      this.#originCity = cities.find((city) => city.coord.is(this.origin)) ?? null;
      this.#destinationCity = cities.find((city) => city.coord.is(this.destination)) ?? null;
    }

    return {
      origin: this.#originCity,
      destination: this.#destinationCity,
    } as const;
  }

  public static create(maneuver: Maneuver) {
    if (maneuver instanceof ManeuverImpl) {
      return maneuver;
    }

    return new ManeuverImpl(maneuver);
  }

  public static async load(id: ManeuverId) {
    const maneuver = await getManeuver(id);
    return maneuver ? ManeuverImpl.create(maneuver) : null;
  }
}
